src/transformers/modeling_utils.py

```diff
@@ -786,145 +786,152 @@ def _load_state_dict_into_meta_model(
-
-                if param_device == "disk" and not is_safetensors:
-                    offload_index = offload_weight(param[:], module_name, offload_folder, offload_index)
-                elif param_device == "cpu" and state_dict_index is not None:
-                    state_dict_index = offload_weight(param[:], module_name, state_dict_folder, state_dict_index)
-                elif (
-                    not is_quantized
-                    or (not hf_quantizer.requires_parameters_quantization)
-                    or (
-                        not hf_quantizer.check_quantized_param(
-                            model, param, module_name, state_dict, param_device=param_device, device_map=device_map
-                        )
-                    )
-                ):
-                    if is_fsdp_enabled():
-                        param_device = "cpu" if is_local_dist_rank_0() else "meta"
-                    module = model.get_submodule(layer)
-                    if param_casting_dtype is not None and param_casting_dtype != empty_param.dtype:
-                        param = param[:].to(param_casting_dtype)
-                    module.load_state_dict(
-                        {param_type: param[:].to(param_device)},
-                        False,
-                        True,
-                    )
-                else:
-                    hf_quantizer.create_quantized_param(
-                        model, param[:], module_name, param_device, state_dict, unexpected_keys
+                    param = param[rank * (row // device_mesh.size()) : (rank + 1) * (row // device_mesh.size()), :]
+                    shard = Shard(0)
+                if param_casting_dtype is not None and param_casting_dtype != empty_param.dtype:
+                    param = param.to(param_casting_dtype)
+                local_parameter = DTensor.from_local(
+                    param,
+                    device_mesh=device_mesh,
+                    placements=[shard] * device_mesh.ndim,
+                )
+                if isinstance(module_to_tp.weight, nn.Parameter):
+                    local_parameter = torch.nn.Parameter(local_parameter)
+                module_to_tp.weight = local_parameter
+                input_fn = partial(tp_layer._prepare_input_fn, tp_layer.input_layouts, tp_layer.desired_input_layouts)
+                output_fn = partial(tp_layer._prepare_output_fn, tp_layer.output_layouts, tp_layer.use_local_output)
+                distribute_module(module_to_tp, device_mesh, None, input_fn, output_fn)
+            else:
+                module_to_tp.load_state_dict({param_type: param[:]}, strict=False, assign=True)
+
+        else:
+            if device_map is None:
+                param_device = "cpu"
+            else:
+                module_name = fixed_param_name
+                while len(module_name) > 0 and module_name not in device_map:
+                    module_name = ".".join(module_name.split(".")[:-1])
```

Comment:
that's not the correct way to fix it 😉 
there was a problem with the regex for sure, but this is just too inneficient. 
```suggestion
                    module_name = module_name.rsplit(".", 1)[0]
                    device_map_regex = "|".join(list(device_map.keys())[::-1])
                    module_layer = re.search(device_map_regex, module_name)
                    if module_name == "" or device_map_regex is None:
                        raise ValueError(
                            f"`device_map` is used, but {module_name} doesn't have any device set. {device_map}"
                        )
                    else:
                        param_device = device_map[module_layer.group()]

```
longest first search! also checking if "." in layername?

Reply:
sounds good. I optimized a little bit more and tested it